src/librustdoc/clean/types.rs

```diff
@@ -1372,56 +1372,61 @@ crate struct PolyTrait {
     crate generic_params: Vec<GenericParamDef>,
 }
 
-/// A representation of a type suitable for hyperlinking purposes. Ideally, one can get the original
-/// type out of the AST/`TyCtxt` given one of these, if more information is needed. Most
-/// importantly, it does not preserve mutability or boxes.
+/// Rustdoc's representation of types, mostly based on the [`hir::Ty`].
 #[derive(Clone, PartialEq, Eq, Debug, Hash)]
 crate enum Type {
-    /// Structs/enums/traits (most that would be an `hir::TyKind::Path`).
-    ResolvedPath {
-        path: Path,
-        did: DefId,
-    },
-    /// `dyn for<'a> Trait<'a> + Send + 'static`
+    /// A named type, which could be a trait.
+    ///
+    /// This is mostly Rustdoc's version of [`hir::Path`].
+    ResolvedPath { path: Path, did: DefId },
+    /// A `dyn Trait` object: `dyn for<'a> Trait<'a> + Send + 'static`
     DynTrait(Vec<PolyTrait>, Option<Lifetime>),
-    /// For parameterized types, so the consumer of the JSON don't go
-    /// looking for types which don't exist anywhere.
+    /// A type parameter.
     Generic(Symbol),
-    /// Primitives are the fixed-size numeric types (plus int/usize/float), char,
-    /// arrays, slices, and tuples.
+    /// A primitive (aka, builtin) type.
+    ///
+    /// This represents things like `i32`, `str`, etc.
     Primitive(PrimitiveType),
-    /// `extern "ABI" fn`
+    /// A function pointer: `extern "ABI" fn(...) -> ...`
     BareFunction(Box<BareFunctionDecl>),
+    /// A tuple type: `(i32, &str)`.
     Tuple(Vec<Type>),
+    /// A slice type (does *not* include the `&`): `[i32]`
     Slice(Box<Type>),
-    /// The `String` field is about the size or the constant representing the array's length.
+    /// An array type.
+    ///
+    /// The `String` field is a stringified version of the array's length parameter.
     Array(Box<Type>, String),
+    /// The never type: `!`
     Never,
+    /// A raw pointer type: `*const i32`, `*mut i32`
     RawPointer(Mutability, Box<Type>),
-    BorrowedRef {
-        lifetime: Option<Lifetime>,
-        mutability: Mutability,
-        type_: Box<Type>,
-    },
+    /// A reference type: `&i32`, `&'a mut Foo`
+    BorrowedRef { lifetime: Option<Lifetime>, mutability: Mutability, type_: Box<Type> },
 
-    // `<Type as Trait>::Name`
+    /// A qualified path to an associated item: `<Type as Trait>::Name`
     QPath {
         name: Symbol,
         self_type: Box<Type>,
+        /// FIXME: This is a hack that should be removed; see [this discussion][1].
+        ///
+        /// [1]: https://github.com/rust-lang/rust/pull/85479#discussion_r635729093
         self_def_id: Option<DefId>,
         trait_: Path,
     },
 
-    // `_`
+    /// A type that is inferred: `_`
     Infer,
```

Comment:
When can this happen? Shouldn't all types be inferred by the time rustdoc gets to them?

The only uses I see are
```rust
src/librustdoc/clean/mod.rs
1295:                // Rustdoc handles `ty::Error`s by turning them into `Type::Infer`s.
1296:                ty::Error(_) => return Type::Infer,
1362:            // Rustdoc handles `TyKind::Err`s by turning them into `Type::Infer`s.
1363:            TyKind::Infer | TyKind::Err => Infer,
1758:                        hir::GenericArg::Infer(_inf) => GenericArg::Infer,
```
so maybe this should say "a type error" instead and be removed? Is it possible to get `TyKind::Infer` at all this late in analysis?

Reply:
> When can this happen? Shouldn't all types be inferred by the time rustdoc gets to them?

I don't know. Probably yes, since we don't even look at item bodies.

> so maybe this should say "a type error" instead and be removed?

What do you mean by "this should \[...\] be removed"? Either way, let's make this change in a separate PR so this one doesn't grow :)

Reply:
I mean remove the `InferTy` variant of GenericArg altogether.

Yeah, another PR sounds fine.